salt/cli/batch_async.py

```diff
@@ -0,0 +1,174 @@
+# -*- coding: utf-8 -*-
+'''
+Execute batch runs
+'''
+
+# Import python libs
+from __future__ import absolute_import, print_function, unicode_literals
+import tornado
+
+# Import salt libs
+import salt.client
+
+# pylint: enable=import-error,no-name-in-module,redefined-builtin
+import logging
+import fnmatch
+
+log = logging.getLogger(__name__)
+
+from salt.cli.batch import _get_bnum, _batch_get_opts, _batch_get_eauth
+
+
+class BatchAsync(object):
+    '''
+    Manage the execution of batch runs
+    '''
+    def __init__(self, parent_opts, jid_gen, clear_load):
+        ioloop = tornado.ioloop.IOLoop.current()
+        self.local = salt.client.get_local_client(parent_opts['conf_file'])
+        clear_load['gather_job_timeout'] = clear_load['kwargs'].pop('gather_job_timeout')
+        self.opts = _batch_get_opts(
+            clear_load.pop('tgt'),
+            clear_load.pop('fun'),
+            clear_load['kwargs'].pop('batch'),
+            self.local.opts,
+            **clear_load)
+        self.eauth = _batch_get_eauth(clear_load['kwargs'])
+        self.minions = set()
+        self.down_minions = set()
+        self.timedout_minions = set()
+        self.done_minions = set()
+        self.to_run = set()
+        self.active = set()
+        self.initialized = False
+        self.ping_jid = jid_gen()
+        self.batch_jid = jid_gen()
+        self.find_job_jid = jid_gen()
+        self.find_job_returned = set()
+        self.event = salt.utils.event.get_event(
+            'master',
+            self.opts['sock_dir'],
+            self.opts['transport'],
+            opts=self.opts,
+            listen=True,
+            io_loop=ioloop,
+            keep_loop=True)
+        self.__set_event_handler()
+
+    def __set_event_handler(self):
+        ping_return_pattern = 'salt/job/{0}/ret/*'.format(self.ping_jid)
+        batch_return_pattern = 'salt/job/{0}/ret/*'.format(self.batch_jid)
+        find_job_return_pattern = 'salt/job/{0}/ret/*'.format(self.find_job_jid)
+        self.event.subscribe(ping_return_pattern, match_type='glob')
+        self.event.subscribe(batch_return_pattern, match_type='glob')
+        self.event.patterns = {
+            (ping_return_pattern, 'ping_return'),
+            (batch_return_pattern, 'batch_run'),
+            (find_job_return_pattern, 'find_job_return')
+        }
+        if not self.event.subscriber.connected():
+            # TODO is there a way to subscribe to only some tags?
```

Comment:
@moio yes, but the implementation is like a filtering after the event handler is called. What I had in mind, but I could not find a way to do it was to register the event handler only for some patterns so that the filtering would happen before the event handler would even be called.

Reply:
I would say, unless some upstream reviewer can indicate a way, you could just remove this TODO then. Thanks for the explanation.